torch/nn/modules/transformer.py

```diff
@@ -55,26 +55,39 @@ def __init__(self, d_model=512, nhead=8, num_encoder_layers=6,
         self.d_model = d_model
         self.nhead = nhead
 
-    def forward(self, src, tgt, src_mask=None, tgt_mask=None, memory_mask=None):
+    def forward(self, src, tgt, src_mask=None, tgt_mask=None,
+                memory_mask=None, src_key_padding_mask=None,
+                tgt_key_padding_mask=None, memory_key_padding_mask=None):
         r"""Take in and process masked source/target sequences.
 
         Args:
             src: the sequence to the encoder (required).
             tgt: the sequence to the decoder (required).
-            src_mask: the mask for the src sequence (optional).
-            tgt_mask: the mask for the tgt sequence (optional).
-            memory_mask: the mask for the encoder output (optional).
+            src_mask: the additive mask for the src sequence (optional).
+            tgt_mask: the additive mask for the tgt sequence (optional).
+            memory_mask: the additive mask for the encoder output (optional).
+            src_key_padding_mask: the ByteTensor mask for src keys per batch (optional).
+            tgt_key_padding_mask: the ByteTensor mask for tgt keys per batch (optional).
+            memory_key_padding_mask: the ByteTensor mask for memory keys per batch (optional).
 
         Shape:
             - src: :math:`(S, N, E)`.
             - tgt: :math:`(T, N, E)`.
             - src_mask: :math:`(S, S)`.
             - tgt_mask: :math:`(T, T)`.
             - memory_mask: :math:`(T, S)`.
-
-            Note: The maksed positions are filled with float('-inf'). Unmasked positions
-            are filled with float(0.0). Masks ensure that the predictions 
-            for position i depend only on the information before position i.
+            - src_key_padding_mask: :math:`(N, S)`.
+            - tgt_key_padding_mask: :math:`(N, T)`.
+            - memory_key_padding_mask: :math:`(N, S)`.
```

Comment:
Since memory_key_padding_mask is used on the multihead_attention layer in decoder, it's dimension should be consistent with that of tgt, rather than src. I guess it is just a typo. 
I'm wondering if we should really have a memory_key_padding_mask, unless a standalone TransformerDecoder is used and memory is passed to the decoder. In that situation, I think the answer is yes, right?

Reply:
I don't think that is True. Memory_key_padding_mask is used on the memory, which is the same dimensions as src. You can see in my tests I always use src length for the memory mask.

Reply:
In regard to your second question, memory_key_padding_mask give the user the flexibility to not attend to positions in the memory that were padded on the src input.